### PR TITLE
Updated getting started doc

### PR DESCRIPTION
**Pull request by chiamp**
Updated getting started doc, as part of content restructuring mentioned in #2627. View the doc [here](https://flax--2698.org.readthedocs.build/en/2698/getting_started.html).

Re-named "Getting Started" to "Quick Start"

### CODE REVIEW

File: docs/getting_started.md [hunk 1]
```diff
@@ -30,35 +32,31 @@ If you see any changes between the two feel free to create a
 [pull request](https://github.com/google/flax/compare)
 to synchronize this Colab with the actual example.
 
++++ {"id": "nwJWKIhdwxDo"}
+
+## 1. Install Flax
```
Comment (cgarciae) [inline docs/getting_started.md:37]: Should we remove the numbering (1, 2, ...)? Doesn't add much to the document IMO but makes it annoying to maintain.

File: docs/getting_started.md [hunk 2]
```diff
@@ -104,63 +126,40 @@ class CNN(nn.Module):
...
-  ds_builder = tfds.builder('mnist')
-  ds_builder.download_and_prepare()
-  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
-  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
-  train_ds['image'] = jnp.float32(train_ds['image']) / 255.
-  test_ds['image'] = jnp.float32(test_ds['image']) / 255.
-  return train_ds, test_ds
+def cross_entropy_loss(*, logits, labels):
+  labels_onehot = jax.nn.one_hot(labels, num_classes=10)
+  return optax.softmax_cross_entropy(logits=logits, labels=labels_onehot).mean()
```
Comment (cgarciae) [inline docs/getting_started.md:159]:
```suggestion
  return optax.softmax_cross_entropy_with_integer_labels(
    logits=logits, labels=labels).mean()
```

File: docs/getting_started.md (hunk 1 above)
Comment (8bitmp3) [reply docs/getting_started.md:37]: IMO, it helps follow all the steps especially for new Flax users.

File: docs/getting_started.md [hunk 3]
```diff
@@ -13,14 +13,16 @@ kernelspec:
...
 [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/flax/blob/main/docs/getting_started.ipynb)
 [![Open On GitHub](https://img.shields.io/badge/Open-on%20GitHub-blue?logo=GitHub)](https://github.com/google/flax/blob/main/docs/getting_started.ipynb)
 
 # Getting Started
 
-This tutorial demonstrates how to construct a simple convolutional neural
+Welcome to Flax!
+
+Flax is an open source Python neural network library that's built on top of [JAX](https://github.com/google/jax). This tutorial demonstrates how to construct a simple convolutional neural
 network (CNN) using the [Flax](https://flax.readthedocs.io) Linen API and train
```
Comment (8bitmp3) [inline docs/getting_started.md:26]: Nit: "and train it..."

File: docs/getting_started.md [hunk 4]
```diff
@@ -13,14 +13,16 @@ kernelspec:
...
 
 [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/flax/blob/main/docs/getting_started.ipynb)
 [![Open On GitHub](https://img.shields.io/badge/Open-on%20GitHub-blue?logo=GitHub)](https://github.com/google/flax/blob/main/docs/getting_started.ipynb)
 
 # Getting Started
 
-This tutorial demonstrates how to construct a simple convolutional neural
+Welcome to Flax!
+
+Flax is an open source Python neural network library that's built on top of [JAX](https://github.com/google/jax). This tutorial demonstrates how to construct a simple convolutional neural
```
Comment (8bitmp3) [inline docs/getting_started.md:25]: Nit:

```suggestion
Flax is an open source Python neural network library built on top of [JAX](https://github.com/google/jax). This tutorial demonstrates how to construct a simple convolutional neural
```

File: docs/getting_started.md [hunk 5]
```diff
@@ -30,62 +32,105 @@ If you see any changes between the two feel free to create a
...
+---
+executionInfo:
+  elapsed: 54
+  status: ok
+  timestamp: 1671500846075
+  user:
+    displayName: Marcus Chiam
+    userId: '17531616275590396120'
+  user_tz: 300
+id: a9633134
```
Comment (8bitmp3) [inline docs/getting_started.md:64]: Can assist with cleaning up the Colab Jupyter metadata. @IvyZX may also have a method.

File: docs/getting_started.md [hunk 6]
```diff
@@ -30,62 +32,105 @@ If you see any changes between the two feel free to create a
...
 import jax
 import jax.numpy as jnp                # JAX NumPy
 
 from flax import linen as nn           # The Linen API
 from flax.training import train_state  # Useful dataclass to keep train state
+from flax import struct                # Flax dataclasses
 
 import numpy as np                     # Ordinary NumPy
 import optax                           # Optimizers
+import tensorflow as tf                # Tensorflow to operate on TFDS
```
Comment (8bitmp3) [inline docs/getting_started.md:75]: Nit: "TensorFlow"

Maybe: 

```suggestion
import tensorflow as tf                # TensorFlow for certain ops
```

WDYT?

TF is used for like `tf.random.set_seed(0)` as well as `tf.cast()` and setting dtypes for e.g. `tf.float32`.

TFDS is `tensorflow_datasets` (package), isn't it?

File: docs/getting_started.md [hunk 7]
```diff
@@ -30,62 +32,105 @@ If you see any changes between the two feel free to create a
...
 ```
 
-+++ {"id": "d0FW1DHa6cfH"}
++++ {"id": "7057395a"}
 
-## 2. Define network
+## 4. Define network
 
 Create a convolutional neural network with the Linen API by subclassing
 [Module](https://flax.readthedocs.io/en/latest/flax.linen.html#core-module-abstraction).
```
Comment (8bitmp3) [inline docs/getting_started.md:115]: Nit:

In other guides we are/started using "Flax Module" since "module/Module" is a common word.

Maybe here:

```suggestion
[Flax Module](https://flax.readthedocs.io/en/latest/flax.linen.html#core-module-abstraction).
```

And then repeat this, which can help new users.

File: docs/getting_started.md [hunk 8]
```diff
@@ -104,63 +149,55 @@ class CNN(nn.Module):
...
-+++ {"id": "xDEoAprU6_JZ"}
++++ {"id": "hy7iRu7_zlx-"}
 
-## 3. Define loss
-
-We simply use `optax.softmax_cross_entropy()`. Note that this function expects both `logits` and `labels` to have shape `[batch, num_classes]`. Since the labels will be read from TFDS as integer values, we first need to convert them to a onehot encoding.
+### View model layers
 
-Our function returns a simple scalar value ready for optimization, so we first take the mean of the vector shaped `[batch]` returned by Optax's loss function.
+Create an instance of the Module and use the [`Module.tabulate`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.tabulate) method to visualize a table of the model layers by passing an RNG key and template image input.
```
Comment (8bitmp3) [inline docs/getting_started.md:156]: As before:

```suggestion
Create an instance of the Flax Module and use the [`Module.tabulate`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.tabulate) method to visualize a table of the model layers by passing an RNG key and template image input.
```

File: docs/getting_started.md [hunk 9]
```diff
@@ -104,63 +149,55 @@ class CNN(nn.Module):
...
-      'loss': loss,
-      'accuracy': accuracy,
-  }
-  return metrics
-```
-
-+++ {"id": "lYz0Emry-ele"}
-
-## 5. Loading data
+We simply use `optax.softmax_cross_entropy()`. Note that this function expects both `logits` and `labels` to have shape `[batch, num_classes]`. Since the labels will be read from TFDS as integer values, we first need to convert them to a onehot encoding.
```
Comment (8bitmp3) [inline docs/getting_started.md:179]: Nit:
- Add a link to the Optax softmax cross entropy API doc and mention Optax since it's an external library.
- If you can, use "second person" like "you"/"your" (Google Style Guide).

For example:

```suggestion
For your loss, use a predefined [`optax.softmax_cross_entropy()`](https://optax.readthedocs.io/en/latest/api.html#optax.softmax_cross_entropy) from the Optax library. Note that this function expects both `logits` and `labels` to have shape `[batch, num_classes]`. Since the labels will be read from TFDS as integer values, first convert them to a one-hot encoding.
```

File: docs/getting_started.md [hunk 10]
```diff
@@ -104,63 +149,55 @@ class CNN(nn.Module):
...
+---
+def cross_entropy_loss(*, logits, labels):
+  labels_onehot = jax.nn.one_hot(labels, num_classes=10)
+  return optax.softmax_cross_entropy(logits=logits, labels=labels_onehot).mean()
 ```
 
-+++ {"id": "UMFK51rsAUX4"}
++++ {"id": "4b5ac16e"}
 
 ## 6. Create train state
```
Comment (8bitmp3) [inline docs/getting_started.md:202]: Nit: Since it's a Flax term/class maybe use

```suggestion
## 6. Create a `TrainState`
```

File: docs/getting_started.md [hunk 11]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
 to pass around a single argument to functions like `train_step()` (see below).
 
 Because this is such a common pattern, Flax provides the class
-[flax.training.train_state.TrainState](https://flax.readthedocs.io/en/latest/flax.training.html#train-state)
-that serves most basic usecases. Usually one would subclass it to add more data
-to be tracked, but in this example we can use it without any modifications.
+[`flax.training.train_state.TrainState`](https://flax.readthedocs.io/en/latest/flax.training.html#train-state)
+that serves most basic usecases. We can then subclass `TrainState` so that it also contains metrics.
```
Comment (8bitmp3) [inline docs/getting_started.md:212]: Nit:

```suggestion
that serves most basic usecases. You can then subclass `TrainState` so that it also contains metrics.
```

File: docs/getting_started.md [hunk 12]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+  status: ok
+  timestamp: 1671500847489
+  user:
+    displayName: Marcus Chiam
+    userId: '17531616275590396120'
+  user_tz: 300
+id: e0102447
+---
+def create_train_state(module, rng, learning_rate, momentum):
   """Creates initial `TrainState`."""
```
Comment (8bitmp3) [inline docs/getting_started.md:258]: Nit:

```suggestion
  """Creates an initial `TrainState`."""
```

File: docs/getting_started.md [hunk 13]
```diff
@@ -30,62 +32,105 @@ If you see any changes between the two feel free to create a
...
+---
 import jax
 import jax.numpy as jnp                # JAX NumPy
 
 from flax import linen as nn           # The Linen API
 from flax.training import train_state  # Useful dataclass to keep train state
+from flax import struct                # Flax dataclasses
 
 import numpy as np                     # Ordinary NumPy
 import optax                           # Optimizers
```
Comment (8bitmp3) [inline docs/getting_started.md:74]:
```suggestion
import optax                           # Optax for common losses and optimizers
```

File: docs/getting_started.md [hunk 14]
```diff
@@ -104,63 +149,55 @@ class CNN(nn.Module):
...
-```
-
-+++ {"id": "lYz0Emry-ele"}
-
-## 5. Loading data
+We simply use `optax.softmax_cross_entropy()`. Note that this function expects both `logits` and `labels` to have shape `[batch, num_classes]`. Since the labels will be read from TFDS as integer values, we first need to convert them to a onehot encoding.
 
-Define a function that loads and prepares the MNIST dataset and converts the
-samples to floating-point numbers.
+Our function returns a simple scalar value ready for optimization, so we first take the mean of the vector shaped `[batch]` returned by Optax's loss function.
```
Comment (8bitmp3) [inline docs/getting_started.md:181]: Nit:

- If you can, use "second person" like "you"/"your" (Google Style Guide).

```suggestion
Your function returns a simple scalar value ready for optimization, make sure to first take the mean of the vector shaped `[batch]` returned by Optax's loss function.
```

File: docs/getting_started.md [hunk 15]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+  user_tz: 300
+id: e4f6f4d3
+---
+init_rng = jax.random.PRNGKey(0)
 ```
 
-+++ {"id": "mHQi20yVCsSf"}
++++ {"id": "80fbb60b"}
+
+## 11. Initialize train state
```
Comment (8bitmp3) [inline docs/getting_started.md:404]: As mentioned before, maybe:

```suggestion
## 11. Initialize the `TrainState`
```

File: docs/getting_started.md [hunk 16]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+## 10. Seed randomness
 
-Create a model evaluation function that:
-
-- Retrieves the evaluation metrics from the device with `jax.device_get`.
-- Copies the metrics
-  [data stored](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html#how-are-parameters-represented-and-how-do-we-handle-general-differentiable-algorithms-that-update-stateful-variables)
-  in a JAX
-  [pytree](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions).
+- Set TF random seed to ensure dataset shuffling is reproducible.
```
Comment (8bitmp3) [inline docs/getting_started.md:364]: Nit:

```suggestion
- Set the TF random seed to ensure dataset shuffling (with `tf.data.Dataset.shuffle`) is reproducible.
```

File: docs/getting_started.md [hunk 17]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+ax2.legend()
+plt.show()
+plt.clf()
 ```
 
-+++ {"id": "UqNrWu7kIC9S"}
++++ {"id": "qQbKS0tV3sZ1"}
 
-## 14. Train and evaluate
+## 14. Inference on test set
```
Comment (8bitmp3) [inline docs/getting_started.md:583]: Nit:

```suggestion
## 14. Perform inference on the test set
```

File: docs/getting_started.md [hunk 18]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+show_img_grid(
+    [test_batch['image'][idx] for idx in range(25)],
+    [f'label={pred[idx]}' for idx in range(25)],
+)
 ```
 
-+++ {"id": "oKcRiQ89xQkF"}
++++ {"id": "edb528b6"}
 
 Congrats! You made it to the end of the annotated MNIST example. You can revisit
```
Comment (8bitmp3) [inline docs/getting_started.md:659]: Nit:

```suggestion
Congratulations! You made it to the end of the annotated MNIST example. You can revisit
```

Congrats may be considered as slang (https://developers.google.com/style/translation#be-inclusive)

**Review by 8bitmp3 — COMMENTED** [review]
Left a few minor suggestions. Feel free to add them/change them or ignore them 👍  Hope this helps!

File: docs/getting_started.md [hunk 19]
```diff
@@ -30,62 +32,105 @@ If you see any changes between the two feel free to create a
 [pull request](https://github.com/google/flax/compare)
```
Comment (cgarciae) [inline docs/getting_started.md:32]: This is no longer true as we've heavily modified the notebook. Maybe we should remove this note?

File: docs/getting_started.md [hunk 20]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+    displayName: Marcus Chiam
+    userId: '17531616275590396120'
+  user_tz: 300
+id: DFwxgBQf44ks
+---
+for test_batch in test_ds.as_numpy_iterator():
+  test_state = compute_metrics(state=test_state, batch=test_batch)
+  pred = state.apply_fn({'params': state.params}, test_batch['image']) # model inference
+  break # get only the first batch
+pred = pred.argmax(axis=1) # argmax the logits to get predicted labels
```
Comment (cgarciae) [inline docs/getting_started.md:603]: Realistically we might want to create jitted function for inference e.g:
```suggestion
@jax.jit
def pred_step(state, batch):
  logits = state.apply_fn({'params': state.params}, test_batch['image'])
  return logits.argmax(axis=1) 

test_batch = test_ds.as_numpy_iterator().next()
pred = pred_step(state, test_batch)
```

Comment (cgarciae) [reply docs/getting_started.md:603]: Even if we don't create `pred_step` its better to use:
```python
test_batch = test_ds.as_numpy_iterator().next()
```
instead of `break` in the loop.

File: docs/getting_started.md [hunk 21]
```diff
@@ -171,233 +208,453 @@ Also adding optimizer & model to this state has the advantage that we only need
...
+    displayName: Marcus Chiam
+    userId: '17531616275590396120'
+  user_tz: 300
+id: 5d5nF3u44JFI
+outputId: 22e013f6-b9b7-4088-84f3-caaf248377ce
+---
+show_img_grid(
+    [test_batch['image'][idx] for idx in range(25)],
+    [f'label={pred[idx]}' for idx in range(25)],
+)
```
Comment (cgarciae) [inline docs/getting_started.md:654]: All of this can be reduced to:

```python
fig, axs = plt.subplots(5, 5, figsize=(12, 12))
for i, ax in enumerate(axs.flatten()):
    ax.imshow(test_batch['image'][i, ..., 0], cmap='gray')
    ax.set_title(f"label={pred['label'][i]}")
    ax.axis('off')
```

**Comment by chiamp** [comment]
Thanks for the suggestions @cgarciae @8bitmp3! I made some updated changes.

**Review by cgarciae — APPROVED** [review]
Awesome @chiamp, looks very good! Approved.

File: docs/getting_started.md [hunk 22]
```diff
@@ -8,83 +8,103 @@ jupytext:
...
 ---
 
-+++ {"id": "9q9JZjt16AMf"}
++++ {"id": "6eea21b3"}
 
 [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/flax/blob/main/docs/getting_started.ipynb)
 [![Open On GitHub](https://img.shields.io/badge/Open-on%20GitHub-blue?logo=GitHub)](https://github.com/google/flax/blob/main/docs/getting_started.ipynb)
 
-# Getting Started
+# Quick Start
```
Comment (8bitmp3) [inline docs/getting_started.md:21]: Nit: Both "quickstart" and "quick start" seem to be OK. To be consistent with "JAX Quickstart", would it make sense to name our doc "Quickstart" (one word) or "Flax quickstart"?

File: docs/getting_started.md [hunk 23]
```diff
@@ -8,83 +8,103 @@ jupytext:
...
 
 !pip install -q flax
 ```
 
++++ {"id": "b529fbef"}
+
+## 2. Loading data
+
+Flax can use any
+data-loading pipeline and this example demonstrates how to utilize TFDS. Define a function that loads and prepares the MNIST dataset and converts the
```
Comment (8bitmp3) [inline docs/getting_started.md:45]: Nit:

```suggestion
data-loading pipeline and this example demonstrates how to utilize TensorFlow Datasets (TFDS). Define a function that loads and prepares the MNIST dataset and converts the
```

Since we haven't mentioned TFDS before, it may help to spell out the full name of the library

**Review by 8bitmp3 — APPROVED** [review]
